pkg/endpoint/bpf.go

```diff
@@ -1249,14 +1371,111 @@ func (e *Endpoint) applyPolicyMapChanges() error {
 	return nil
 }
 
+// endpointPolicyLockdown puts the endpoint policy map into a lockdown
+// mode. If the policy map can still accommodate all the deny entries (and
+// there is at least one deny policy) then all deny entries are inserted without
+// any allows. If there is no space for the deny entries or there are no deny entries
+// to insert then a new bpf policy map is populated with deny all traffic entries.
+// The old map is unpinned, and the new map is pinned.
+func (e *Endpoint) endpointPolicyLockdown() error {
+	dL := e.desiredPolicy.GetPolicyMap().DenyLen()
+	fullLockdown := dL > policymap.MaxEntries || dL == 0
+
+	// If a full lock down has already happened and that
+	// is what is still required then we do not have to
+	// do anything. If a deny-preserving (or "soft")
+	// lockdown is what is required then we still need to
+	// create a new map.
+	if fullLockdown && e.getState() == StateFullLockdown {
+		return nil
+	}
+
+	var state State
+	tmpPMPath := e.policyMapPath() + "_lockdown"
+	pM, err := policymap.CreateUnpinned(tmpPMPath)
+	if err != nil {
+		return fmt.Errorf("could not create bpf map %q: %w", tmpPMPath, err)
+	}
+	mapState := policy.NewMapState(nil)
+	// Deny entries also exceed max entries, or there are no deny entires,
+	// go into a full lockdown.
+	if fullLockdown {
+		state = StateFullLockdown
+		for _, k := range allTrafficKeys {
+			policymapKey := policymap.NewKey(k.Identity, k.DestPort,
+				k.Nexthdr, k.TrafficDirection)
+			if err := pM.DenyKey(policymapKey); err != nil {
+				return fmt.Errorf("failed to add deny all policy (%v): %w", policymapKey, err)
+			}
+			mapState.Insert(k, policy.MapStateEntry{IsDeny: true})
+		}
+	} else {
+		var outErr error
+		state = StateSoftLockdown
+		e.desiredPolicy.GetPolicyMap().ForEachDeny(func(k policy.Key, entry policy.MapStateEntry) bool {
+			policymapKey := policymap.NewKey(k.Identity, k.DestPort,
+				k.Nexthdr, k.TrafficDirection)
+			err := pM.DenyKey(policymapKey)
+			if err != nil {
+				outErr = err
+				return false
+			}
+			mapState.Insert(k, entry)
+			return true
+		})
+		if outErr != nil {
+			return fmt.Errorf("failed to add all deny policy map entries: %w", outErr)
+		}
+	}
+	// 1. Unpin the old policy.
+	// 2. Set the policy map to the one created in this
+	//    method.
+	// 3. Set the realizedPolicy field to a new EndpointPolicy,
+	//    so that we are not modifying desiredPolicy.
+	// 4. Pin the new bpf map.
+	// 5. Update the pressure metric, because other code paths
+	//    to updating it are now disabled in lockdown mode.
+	// 6. Update the state of the endpoint
+	err = e.policyMap.Unpin()
+	if err != nil {
+		return fmt.Errorf("failed to unpin previous map: %w", err)
+	}
+	e.policyMap = pM
+	e.realizedPolicy = policy.NewEndpointPolicy(e.policyGetter.GetPolicyRepository())
+	e.realizedPolicy.SetPolicyMap(mapState)
+	if err = e.policyMap.Pin(e.policyMapPath()); err != nil {
```

Comment:
The datapath BPF program references a policymap by the fd of the policymap at the time of the BPF programs being loaded.

What this means is that if we create the new policymap and pin it to the filesystem, we have not actually updated the datapath yet, we've just prepared everything in place so that the datapath can now reference the newly created map from the filesystem. The datapath continues to use the old map until the BPF programs are reloaded.

It looks to me like there are two potential callers that come down into this path:
1. `regenerateBPF()` first reloads the BPF programs, then syncs the policymap which calls here. By creating a new policymap at this point in the process, it will not take effect until the next regeneration.
2. `endpointmanager.UpdatePolicyMaps()`. There is no BPF program reload from this path, so the changes would not take effect.